### ISSUE TITLE
Including relationship data when fetching data (aka side loading)

### ISSUE THREAD

**Issue by digitalsadhu** [digitalsadhu/loopback-component-jsonapi#38]
@listepo @tsteuwer 
Below is a write up of my understanding of what we need to do to support returning relationship information in GET requests to /:collection and /:collection/:id

Does it look right?

---------

Loopback allows you to control when relationship data is returned with a url filter:

Eg. 
- GET /:collection?filter={"include": "author"}
- GET /:collection/:id?filter={"include": "author"}

Looking at the JSON API spec: 

>If an endpoint supports the include parameter and a client supplies it, the server MUST NOT include unrequested resource objects in the included section of the compound document.

>The value of the include parameter MUST be a comma-separated (U+002C COMMA, ",") list of relationship paths. A relationship path is a dot-separated (U+002E FULL-STOP, ".") list of relationship names.

>If a server is unable to identify a relationship path or does not support inclusion of resources from a path, it MUST respond with 400 Bad Request.

the ideal would be something like: 

Eg. no include filter in url
GET http://localhost:3000/api/posts/9
```js
{
  "data": {
    "type": "posts",
    "id": "9",
    "attributes": {
      "title": "test post",
      "content": "ttseas"
    },
    "relationships": {
      "author": {
        "links": {
          "related": "http://localhost:3000/posts/9/author"
        }
      }
    }
    "links": {
      "self": "http://localhost:3000/api/posts/9"
    }
  },
  "links": {
    "self": "http://localhost:3000/api/posts/9"
  }
}
```

Eg. include filter in url
GET http://localhost:3000/api/posts/9?filter={"include":"author"}
```js
{
  "data": {
    "type": "posts",
    "id": "9",
    "attributes": {
      "title": "test post",
      "content": "ttseas"
    },
    "relationships": {
      "author": {
        "links": {
          "related": "http://localhost:3000/posts/9/author"
        },
        "data": { "type": "people", "id": "1" }
      }
    }
    "links": {
      "self": "http://localhost:3000/api/posts/9"
    }
  },
  "links": {
    "self": "http://localhost:3000/api/posts/9?filter=%7B%22include%22%3A%20%22author%22%7D"
  },
  "included": [{
    "type": "people",
    "id": "1",
    "attributes": {
      "name": "Bob Jones"
    },
    "links": {
      "self": "http://localhost:3000/people/1"
    }
  }]
}
```

- [x] add thorough test coverage for items below:
- [x] include relationships link in GET requests for collections and resources eg.  "author": { "links": { "related": "http://localhost:3000/posts/9/author" }}
- [x] rewrite url &include=author -> filter={"include":"author"} to support spec requirement
- [x] return 400 error if non existent relationship is included in url eg. &include=idontexist
- [x] remove FK links out of any `attributes` objects. Probably only applies to belongsTo relationship

**Comment by tsteuwer** [comment]
According to the JSON API spec:

[Fetching Includes](http://jsonapi.org/format/#fetching-includes)
> An endpoint MAY also support an include request parameter to allow the client to customize which related resources should be returned.
> [...]
>> GET /articles/1?include=comments HTTP/1.1
>> Accept: application/vnd.api+json

So, in regards to your comment above, you're request would actually look like:

```
GET http://localhost:3000/api/posts/9?include=author
```

Now, technically, we could also filter with this, so that would be:

```
GET http://localhost:3000/api/posts/9?include=author&filter={"firstName":"John"}
```

**Comment by tsteuwer** [comment]
@digitalsadhu I have started on this and will most likley need help with it. I've started the branch ``feature/request-includes``. I have added:

* Check if the query object contains ``includes``.
* Check if the current method supports the parameter (e.g. DELETE, POST) should not include this param.
* Added new test for DELETE if it's included.

... more to come.

**Comment by digitalsadhu** [comment]
:+1 sure thing. Just Sing out.

**Comment by tsteuwer** [comment]
@digitalsadhu delete the branch ``feature/request-includes``, do a fetch, then check out the branch. I've changed a number of things. Here's my concern right now:

While I'm deleting the relationship from the ``attributes`` property, I have to then add it back as either an array of id's (IF it's a hasMany relationship), or just as a single property for a hasOne relationship. How would I find out what the key is for the relationship?

For instance:
**Has One Relationship**
```
data: {
   type: 'post',
   id: 5,
   attributes: {
     content: 'blah',
     other: 'other stuff'
     authorId: 22
   }
},
included: [
   {
      type: 'author',
      authorId: 22
      attributes: {
        firstName: 'John',
        lastName: 'Doe'
      }
   }
]
```

And for a hasMany...
```
data: {
   type: 'reviewerGroup',
   id: 1,
   attributes: {
     title: 'Reviewer Group 1',
     authority: 'Bosses'
     authorId: [22, 53]
   }
},
included: [
   {
      type: 'author',
      authorId: 22
      attributes: {
        firstName: 'John',
        lastName: 'Doe'
      }
   },
   {
      type: 'author',
      authorId: 53
      attributes: {
        firstName: 'Joe',
        lastName: 'Shmoe'
      }
   }
]
``

**Comment by tsteuwer** [comment]
Ok I've figured it out. Have single relationship requests via ``include=author``. Now I need to get multiple relationship requests going, a la ``include=author,posts``

**Comment by tsteuwer** [comment]
@digitalsadhu @BenjaminHorn can I get one of you to CR what I've done so far. Is this the right way to go about this?

**Comment by digitalsadhu** [comment]
Cheers man, will try to take a look today/tonight

**Comment by digitalsadhu** [comment]
@tsteuwer you can always figure out everything you need to about relationships via 
Model.relations for each model. 

So if you are dealing with model Contract then just grab the Contract.relations object and look through it. It has the relationship type (belongsTo, hasOne, hasMany etc), the name of the foreign key and the value for it. It also gives you the from model and the to model but its important to note that the order is reversed for belongsTo. 

Pretty easy to handle though, you just check the relation.type === 'belongsto' and then switch modelTo and modelFrom as @BenjaminHorn has done in his PR. Have a look in there for inspiration if you need to.

**Comment by tsteuwer** [comment]
@digitalsadhu  I did it! wooo hoo :dancers: However, I can not, for the life of me, get 2 spaces working with vim. Is there any way to run a fix on the linter to autho-indent 2 spaces for me?

**Comment by digitalsadhu** [comment]
Nice one!

Haha damn indentation! There is an eslint formatted I believe so it's possible we could auto format yes. I haven't tried it but if you want you could give it a go and if it works ok add it to the project.

**Comment by tsteuwer** [comment]
For se reason the one that comes with the project doesn't work. I've used jscs and it just works. Have you or @benjaminhorn had a chance to check out the cr?

**Comment by digitalsadhu** [comment]
@tsteuwer Ok its Saturday night and I'm ready for open source! Will look through your code now!

**Comment by tsteuwer** [comment]
@digitalsadhu was there anything else we need? Did you look over the code review? I can't remember the last thing that was said about this ticket

**Comment by digitalsadhu** [comment]
Merged into master as v0.11.0
Yay!
